fdbserver/Restore.actor.cpp

```diff
@@ -18,86 +18,604 @@
+				when ( RestoreSimpleRequest req = waitNext(interf.heartbeat.getFuture()) ) {
+					requestTypeStr = "heartbeat";
+					actors.add( handleHeartbeat(req, interf.id()) );
+				}
+				when ( RestoreRecruitRoleRequest req = waitNext(interf.recruitRole.getFuture()) ) {
+					requestTypeStr = "recruitRole";
+					actors.add( handleRecruitRoleRequest(req, self, &actors, cx) );
+				}
+				when ( RestoreSimpleRequest req = waitNext(interf.terminateWorker.getFuture()) ) {
+					// Destroy the worker at the end of the restore
+					// TODO: Cancel its own actors
+					requestTypeStr = "terminateWorker";
+					wait( handlerTerminateWorkerRequest(req, self, interf, cx) );
+					return Void();
+				}
+			}
+
+		} catch (Error &e) {
+			fprintf(stdout, "[ERROR] Loader handle received request:%s error. error code:%d, error message:%s\n",
+					requestTypeStr.c_str(), e.code(), e.what());
+			if ( requestTypeStr.find("[Init]") != std::string::npos ) {
+				printf("Exit due to error at requestType:%s", requestTypeStr.c_str());
+				break;
+			}
+		}
+	}
+
+	return Void();
+}
+
+ACTOR Future<Void> _restoreWorker(Database cx_input, LocalityData locality) {
+	state Database cx = cx_input;
+	state RestoreWorkerInterface workerInterf;
+	workerInterf.initEndpoints();
+	state Optional<RestoreWorkerInterface> leaderInterf;
+	//Global data for the worker
+	state Reference<RestoreWorkerData> self = Reference<RestoreWorkerData>(new RestoreWorkerData());
+
+	self->workerID = workerInterf.id();
+
+	initRestoreWorkerConfig(); //TODO: Change to a global struct to store the restore configuration
+
+	// Compete in registering its restoreInterface as the leader.
+	state Transaction tr(cx);
+	loop {
+		try {
+			tr.reset();
+			tr.setOption(FDBTransactionOptions::ACCESS_SYSTEM_KEYS);
+			tr.setOption(FDBTransactionOptions::LOCK_AWARE);
+			Optional<Value> leader = wait(tr.get(restoreLeaderKey));
+			if(leader.present()) {
+				leaderInterf = BinaryReader::fromStringRef<RestoreWorkerInterface>(leader.get(), IncludeVersion());
+				// NOTE: Handle the situation that the leader's commit of its key causes error(commit_unknown_result)
+				// In this situation, the leader will try to register its key again, which will never succeed.
+				// We should let leader escape from the infinite loop
+				if ( leaderInterf.get().id() == workerInterf.id() ) {
+					printf("[Worker] NodeID:%s is the leader and has registered its key in commit_unknown_result error. Let it set the key again\n",
+							leaderInterf.get().id().toString().c_str());
+					tr.set(restoreLeaderKey, BinaryWriter::toValue(workerInterf, IncludeVersion()));
+					wait(tr.commit());
+					 // reset leaderInterf to invalid for the leader process
+					 // because a process will not execute leader's logic unless leaderInterf is invalid
+					leaderInterf = Optional<RestoreWorkerInterface>();
+					break;
+				}
+				state Standalone<RangeResultRef> agentValues = wait(tr.getRange(restoreWorkersKeys, CLIENT_KNOBS->TOO_MANY));
```

Comment:
The current implementation assume there is no node failure: the leader never changes. So if we wait on the change of leader, we will never get to execute the restore logic.

When we consider the node failure that causes leader change, it will be great to use the exact suggestion above.

For this PR, I will strip out the leader election logic out as `monitorleader(...)`. I will postpone the handling of the leader change to the next PR.